### ISSUE TITLE
QueryBus with Guzzle/Promises?

### ISSUE THREAD

**Issue by enumag** [prooph/service-bus#174]
Is it possible to somehow use QueryBus with [Guzzle/Promises](https://github.com/guzzle/promises)?

The main reason why I'd like to use Guzzle promises is the synchrionous `wait` method. From the [examples](https://hotexamples.com/examples/prooph.servicebus/QueryBus/dispatch/php-querybus-dispatch-method-examples.html) I found for QueryBus it seems we need to get the result with 

```php
		$promise = $this->queryBus->dispatch($query);

		$result = null;
		$promise->then(
			function ($value) use (&$result) {
				$result = $value;
			}
		);

		return $result;
```

which is quite disgusting imo. If there is some nicer way, please tell me.

Also Guzzle promises have more installations on packagist (by far the most installations among all php promise libraries) and implement Promises/A+ where as React only implements the older Promises/A spec.

**Comment by prolic** [comment]
This would be a BC break and can only be done with new major release. But you can also copy paste the query bus implementation and make your desired adjustments, if you need it.

**Comment by codeliner** [comment]
We hoped to see a promise PSR but async-interop group failed :( If someone has an idea how we can use a promise interface only and make it possible to work with different promise implementations this would be a good reason for a new major version. But switching promise implementation is not an option. Can't you use a react promise to guzzle promise converter?

**Comment by enumag** [comment]
I understand - and agree. I was quite confused that there is no Promise PSR yet to be honest.

> Can't you use a react promise to guzzle promise converter?

Maybe, I don't know enough about promises to try that.

**Comment by prolic** [comment]
We could add an own promise interface and provide different implementations
that aggregate whatever needed. Kind of like httplug does.

**Comment by enumag** [comment]
I talked with a friend who explained to me how promises work. Now I see that the example code I linked in the first post will only work if the promise is a FulfilledPromise. So it doesn't make sense to use promises at all since I don't use any non-blocking API to get the results anyway - shame but PHP database API is not really made for that. So what I'll try to do instead is to write a QueryBus which won't use promises at all.
